repos/keg-core/core/scripts/js/getKegEnvs.js

```diff
@@ -0,0 +1,88 @@
+const { get, isStr } = require('@keg-hub/jsutils')
+const getAppConfig = require('@keg-hub/tap-resolver/src/resolvers/getAppConfig')
+const fs = require('fs')
+const path = require('path')
+
+/**
+ * @param {string} tapPath - path to the current tap's root
+ * @param {string} corePath - path to keg-core's root
+ * @return {Object} - object containing tap and core's app configs and package json objects
+ *  - used to resolve replacement values for envs of form "<tap|core>.<package|config>.<some_property>"
+ */
+const buildReplaceContext = (tapPath, corePath) => {
+  const buildContext = rootPath => ({
+    package: rootPath
+      ? require(path.join(rootPath, 'package.json') )
+      : {},
+    config: rootPath
+      ? getAppConfig(rootPath, false, false)
+      : {}
+  })
+
+  return {
+    core: buildContext(corePath),
+    tap: buildContext(tapPath),
+  }
+}
+
+/**
+ * 
+ * @param {Object} replaceContext - context object produced by `buildReplaceContext`
+ * @return {Object} - object of env entries pulled from tap config and core config, 
+ * dynamically replacing any values of form `<tap|core>.<package|config>.<some_property>`
+ */
+const buildEnvs = replaceContext => {
+  const tapEnvs = get(replaceContext, 'tap.config.keg.envs')
+  const coreEnvs = get(replaceContext, 'core.config.keg.envs')
+
+  const envs = {
+    ...coreEnvs,
+    ...tapEnvs
+  }
+
+  const getFallback = value => 
+    isStr(value) && (value.startsWith('tap.') || value.startsWith('core.'))
+      ? undefined
+      : value
+
+  return Object.entries(envs).reduce(
+    (replacements, [key, value]) => {
+      const replacement = get(replaceContext, value, getFallback(value))
+      return {
```

Comment:
I used to do reduce loops this way, but then I stopped because it's creating a new object on every loop
It's not a huge deal, but is a minor performance improvement.
Now I just set the key / value to the object and return it

Reply:
Man, I really like doing it this way. I guess because it's "immutable", and there's a new object every time. 

But you're right, it is slower, and there's nothing I hate more than a really long build step, so I'd rather not contribute to that! 

Updated